HULKs/hulk
# Supertwix

Merged version of multiple separate PRs because some of them depend on each other and having to chain rebases is a pain.
Also, no one was going to review five different PRs where some of the code was already deprecated.

## Introduced Changes

- Dockable panels using [egui_dock](https://crates.io/crates/egui_dock)
- Ball detection overlay
- Popups don't close on the first interaction anymore
- Better connection status indication

## ToDo / Known Issues

- [ ] Program crashes when closing the last tab


## Ideas for Next Iterations (Not This PR)

- Add more keybindings for navigating and manipulating the panel tree.

## How to Test

- Docking
  - Spawn new tabs by clicking the little `+` icon or pressing `Ctrl`+`t`.
  - Tabs can be dragged around to rearrange them, convert them into panels and back.
  - Changing panel type (Text, Map, Image, etc.) should (only) change the active tab.
- Ball detection overlay
  - Open an image panel and enable it in the overlays menu.
  - Should work the same as in flora
- Popups don't close on the first interaction anymore
  - pretty self explanatory
  - less annoying now when toggling more than one map layer
  - perhaps more annoying when just toggling one as you need one extra click
- Status indication
  - The label next to the "connection intent" checkbox now changes both text and color depending on the connection state.
  - White: No connection intent
  - Yellow: Connecting
  - Green: Connected
  - Red: Failed to connect
